python/ray/tune/experiment/trial.py

```diff
@@ -749,23 +749,46 @@ def pickled_error_file(self):
             return None
         return os.path.join(self.local_path, self.run_metadata.pickled_error_filename)
 
-    def handle_error(self, exc: Optional[Union[TuneError, RayTaskError]] = None):
+    def _handle_restore_error(self, exc: _TuneRestoreError):
+        exc = exc.exc
+        if self.temporary_state.num_restore_failures >= int(
+            os.environ.get("TUNE_RESTORE_RETRY_NUM", 0)
+        ):
+            # Restore was unsuccessful, try again without checkpoint.
+            self.clear_checkpoint()
+            self.run_metadata.num_failures += 1
+        else:
+            self.temporary_state.num_restore_failures += 1
+
+    def _handle_ray_actor_error(self, exc: RayActorError):
+        exc._preempted = True  # TODO(justinvyu): Test the real integration
+        if not exc._preempted:
+            # Only count non-preempted actor errors as failures.
+            self.run_metadata.num_failures += 1
+
+    def _handle_ray_task_error(self, exc: RayTaskError):
+        if isinstance(exc.cause, RayActorError):
+            # Handle the RayActorError directly (ex: Ray Train worker actor errors)
+            return self._handle_ray_actor_error(exc.cause)
+
+        # Increment failures for all user errors (which get raised as RayTaskError)
+        self.run_metadata.num_failures += 1
```

Comment:
Key change 2: this is the actual logic of the PR.

Question: Is it ok to treat `RayTaskError` with a cause of `RayActorError` so broadly like this? One strawman counterexample:

```python
def tune_fn_trainable(config):
    e = RayActorError()
    e._preempted = True
    raise e

tune.Tuner(tune_fn_trainable).fit()
```

Reply:
Another possibility would be to have the `DataParallelTrainer` pass through the pre-emption `RayActorError` as a special case, but I feel like that's more misleading, as it's disguising the coordinator's error with the worker's error.

Reply:
TODO: use `exc.as_instanceof_cause()` instead of the private `cause` attr.

Reply:
This seems okay for now... this seems clean enough for now such that if new use cases come up in the future we can separate this logic and improve it further.

Reply:
TODO: add the configurability of whether or not to count preemption errors here.